zebra-rpc/src/methods.rs

```diff
@@ -2215,18 +2215,110 @@ where
+        if let get_block_template::Response::ProposalMode(response) = validation_result {
+            if !response.is_valid() {
+                tracing::debug!("block template is invalid");
+
+                #[cfg(not(test))]
+                let tx_iter = mempool_txs.iter();
+                #[cfg(test)]
+                let tx_iter = mempool_txs.iter().map(|(_, tx)| tx);
+
+                for tx in tx_iter {
+                    let tx_id = tx.transaction.id;
+
+                    // Filter mempool without the current tx
+                    #[cfg(not(test))]
+                    let trimmed_mempool: Vec<_> = mempool_txs
+                        .iter()
+                        .filter(|t| t.transaction.id != tx_id)
+                        .cloned()
+                        .collect();
+
+                    #[cfg(test)]
+                    let trimmed_mempool: Vec<_> = mempool_txs
+                        .iter()
+                        .filter(|(_, t)| t.transaction.id != tx_id)
+                        .cloned()
+                        .collect();
+
+                    let trimmed_response = GetBlockTemplate::new(
+                        &network,
+                        &miner_address,
+                        &chain_tip_and_local_time,
+                        server_long_poll_id,
+                        trimmed_mempool.clone(),
+                        submit_old,
+                        debug_like_zcashd,
+                        extra_coinbase_data.clone(),
+                    );
+
+                    let block_bytes =
+                        proposal_block_from_template(&trimmed_response, None, network_upgrade)
+                            .map_err(to_jsonrpc_error)?
+                            .zcash_serialize_to_vec()
+                            .map_err(to_jsonrpc_error)?;
+
+                    if let Ok(get_block_template::Response::ProposalMode(response)) =
+                        get_block_template::validate_block_proposal(
+                            self.gbt.block_verifier_router(),
+                            block_bytes,
+                            network.clone(),
+                            latest_chain_tip.clone(),
+                            sync_status.clone(),
+                        )
+                        .await
+                    {
+                        if response.is_valid() {
+                            tracing::debug!(
+                                "block template is valid after removing tx {:?}",
+                                tx_id
+                            );
+
+                            // Remove the transaction from the mempool
+                            let _ = mempool
+                                .clone()
+                                .oneshot(mempool::Request::RemoveTransaction(tx_id))
+                                .await;
+
+                            return Ok(response.into());
```

Comment:
I think this is better than what I had in mind, which was to leave the transactions in the mempool and filter them out in this function.

This code should wait for all of the proposal results, remove transactions in the mempool that were in invalid blocks rather than those in valid blocks, and then call the mempool again to get what's left in the verified set.

It also needs to consider transactions dependencies, otherwise TEX transactions will never pass this validation step.

It might be better to do this in the mempool instead, immediately after adding transactions to the verified set (so it'll still be auto-verified by the transaction verifier when re-submitted as block requests) but before gossiping to peers.